src/Http/Routing/src/RequestDelegateFilterPipelineBuilder.cs

```diff
@@ -0,0 +1,72 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Text.Json.Serialization.Metadata;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Json;
+using Microsoft.AspNetCore.Internal;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Routing;
+
+internal static class RequestDelegateFilterPipelineBuilder
+{
+    // Due to https://github.com/dotnet/aspnetcore/issues/41330 we cannot reference the EmptyHttpResult type
+    // but users still need to assert on it as in https://github.com/dotnet/aspnetcore/issues/45063
+    // so we temporarily work around this here by using reflection to get the actual type.
+    private static readonly object? EmptyHttpResultInstance = Type.GetType("Microsoft.AspNetCore.Http.HttpResults.EmptyHttpResult, Microsoft.AspNetCore.Http.Results")?.GetProperty("Instance")?.GetValue(null, null);
+
+    public static RequestDelegate Create(RequestDelegate requestDelegate, RequestDelegateFactoryOptions options)
+    {
+        Debug.Assert(options.EndpointBuilder != null);
+
+        var serviceProvider = options.ServiceProvider ?? options.EndpointBuilder.ApplicationServices;
+        var jsonOptions = serviceProvider?.GetService<IOptions<JsonOptions>>()?.Value ?? new JsonOptions();
+        var jsonSerializerOptions = jsonOptions.SerializerOptions;
+
+        var factoryContext = new EndpointFilterFactoryContext
+        {
+            MethodInfo = requestDelegate.Method,
+            ApplicationServices = options.EndpointBuilder.ApplicationServices
+        };
+        var jsonTypeInfo = (JsonTypeInfo<object>)jsonSerializerOptions.GetReadOnlyTypeInfo(typeof(object));
```

Comment:
How does this work with source generation, if we always get the JsonTypeInfo for `object`?

Reply:
https://github.com/dotnet/aspnetcore/pull/46020/files#diff-8e7a4b7b1b9fd8d37f7052f449953179d0e66edb4b10ea693a7a0dd18506fec1R51-R52

Reply:
This is weird, but OK I guess.

Reply:
Very weird.